File: lc_py/lcPy/design/designLinkedList.py
# LeetCode 707: 设计链表
# 设计链表的实现。您可以选择使用单链表或双链表。

from typing import Optional, List, Any
from collections import deque
import weakref

class ListNode:
    """双向链表节点"""
    def __init__(self, val: int = 0):
        self.val = val
        self.prev: Optional['ListNode'] = None
        self.next: Optional['ListNode'] = None
    
    def __repr__(self):
        return f"ListNode({self.val})"

class MyLinkedList:
    """双向链表实现"""
    
    def __init__(self):
        """初始化链表，使用哨兵节点"""
        self.head = ListNode()  # 哨兵头节点
        self.tail = ListNode()  # 哨兵尾节点
        self.head.next = self.tail
        self.tail.prev = self.head
        self.size = 0
    
    def get(self, index: int) -> int:
        """获取链表中第 index 个节点的值"""
        if index < 0 or index >= self.size:
            return -1
        
        current = self.head.next
        for _ in range(index):
            current = current.next
        return current.val
    
    def addAtHead(self, val: int) -> None:
        """在链表第一个元素之前添加一个值为 val 的节点"""
        self.addAtIndex(0, val)
    
    def addAtTail(self, val: int) -> None:
        """将值为 val 的节点追加到链表的最后一个元素"""
        self.addAtIndex(self.size, val)
    
    def addAtIndex(self, index: int, val: int) -> None:
        """在链表中的第 index 个节点之前添加值为 val 的节点"""
        if index < 0 or index > self.size:
            return
        
        pred = self.head
        for _ in range(index):
            pred = pred.next
        
        succ = pred.next
        new_node = ListNode(val)
        
        new_node.prev = pred
        new_node.next = succ
        pred.next = new_node
        succ.prev = new_node
        
        self.size += 1
    
    def deleteAtIndex(self, index: int) -> None:
        """如果索引 index 有效，则删除链表中的第 index 个节点"""
        if index < 0 or index >= self.size:
            return
        
        pred = self.head
        for _ in range(index):
            pred = pred.next
        
        to_delete = pred.next
        succ = to_delete.next
        
        pred.next = succ
        succ.prev = pred
        
        self.size -= 1
    
    def __str__(self) -> str:
        """字符串表示"""
        result = []
        current = self.head.next
        while current != self.tail:
            result.append(str(current.val))
            current = current.next
        return f"[{', '.join(result)}]"
    
    def __len__(self) -> int:
        """返回链表长度"""
        return self.size
    
    def to_list(self) -> List[int]:
        """转换为Python列表"""
        result = []
        current = self.head.next
        while current != self.tail:
            result.append(current.val)
            current = current.next
        return result

class ListLinkedList:
    """使用Python列表实现的链表"""
    
    def __init__(self):
        self.data = []
    
    def get(self, index: int) -> int:
        if 0 <= index < len(self.data):
            return self.data[index]
        return -1
    
    def addAtHead(self, val: int) -> None:
        self.data.insert(0, val)
    
    def addAtTail(self, val: int) -> None:
        self.data.append(val)
    
    def addAtIndex(self, index: int, val: int) -> None:
        if 0 <= index <= len(self.data):
            self.data.insert(index, val)
    
    def deleteAtIndex(self, index: int) -> None:
        if 0 <= index < len(self.data):
            self.data.pop(index)
    
    def __str__(self) -> str:
        return str(self.data)

class DequeLinkedList:
    """使用collections.deque实现的链表"""
    
    def __init__(self):
        self.data = deque()
    
    def get(self, index: int) -> int:
        if 0 <= index < len(self.data):
            return self.data[index]
        return -1
    
    def addAtHead(self, val: int) -> None:
        self.data.appendleft(val)
    
    def addAtTail(self, val: int) -> None:
        self.data.append(val)
    
    def addAtIndex(self, index: int, val: int) -> None:
        if 0 <= index <= len(self.data):
            self.data.insert(index, val)
    
    def deleteAtIndex(self, index: int) -> None:
        if 0 <= index < len(self.data):
            del self.data[index]
    
    def __str__(self) -> str:
        return str(list(self.data))

class SinglyListNode:
    """单链表节点"""
    def __init__(self, val: int = 0):
        self.val = val
        self.next: Optional['SinglyListNode'] = None

class SinglyLinkedList:
    """单链表实现"""
    
    def __init__(self):
        self.head = SinglyListNode()  # 哨兵节点
        self.size = 0
    
    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1
        
        current = self.head.next
        for _ in range(index):
            current = current.next
        return current.val
    
    def addAtHead(self, val: int) -> None:
        self.addAtIndex(0, val)
    
    def addAtTail(self, val: int) -> None:
        self.addAtIndex(self.size, val)
    
    def addAtIndex(self, index: int, val: int) -> None:
        if index < 0 or index > self.size:
            return
        
        pred = self.head
        for _ in range(index):
            pred = pred.next
        
        new_node = SinglyListNode(val)
        new_node.next = pred.next
        pred.next = new_node
        
        self.size += 1
    
    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return
        
        pred = self.head
        for _ in range(index):
            pred = pred.next
        
        pred.next = pred.next.next
        self.size -= 1

class FunctionalLinkedList:
    """函数式编程风格的链表"""
    
    def __init__(self):
        self.data = []
    
    def get(self, index: int) -> int:
        return self.data[index] if 0 <= index < len(self.data) else -1
    
    def addAtHead(self, val: int) -> None:
        self.data = [val] + self.data
    
    def addAtTail(self, val: int) -> None:
        self.data = self.data + [val]
    
    def addAtIndex(self, index: int, val: int) -> None:
        if 0 <= index <= len(self.data):
            self.data = self.data[:index] + [val] + self.data[index:]
    
    def deleteAtIndex(self, index: int) -> None:
        if 0 <= index < len(self.data):
            self.data = self.data[:index] + self.data[index + 1:]
    
    def __str__(self) -> str:
        return str(self.data)

class GenericLinkedList:
    """泛型链表实现"""
    
    def __init__(self, data_type: type = int):
        self.data_type = data_type
        self.data = []
    
    def get(self, index: int) -> Any:
        if 0 <= index < len(self.data):
            return self.data[index]
        return None
    
    def add(self, index: int, val: Any) -> None:
        if not isinstance(val, self.data_type):
            raise TypeError(f"Expected {self.data_type}, got {type(val)}")
        
        if 0 <= index <= len(self.data):
            self.data.insert(index, val)
    
    def delete(self, index: int) -> None:
        if 0 <= index < len(self.data):
            self.data.pop(index)
    
    def __str__(self) -> str:
        return f"{self.data_type.__name__}LinkedList{self.data}"

def linked_list_operations(operations: List[tuple]) -> List[Optional[int]]:
    """函数式操作链表"""
    obj = MyLinkedList()
    results = []
    
    for op, *params in operations:
        if op == "get":
            results.append(obj.get(params[0]))
        elif op == "addAtHead":
            obj.addAtHead(params[0])
            results.append(None)
        elif op == "addAtTail":
            obj.addAtTail(params[0])
            results.append(None)
        elif op == "addAtIndex":
            obj.addAtIndex(params[0], params[1])
            results.append(None)
        elif op == "deleteAtIndex":
            obj.deleteAtIndex(params[0])
            results.append(None)
        else:
            results.append(None)
    
    return results

def test_design_linked_list():
    """测试设计链表"""
    print("=== 设计链表测试 ===")
    
    # 测试双向链表实现
    print("\n--- 双向链表实现测试 ---")
    linked_list = MyLinkedList()
    
    linked_list.addAtHead(7)
    linked_list.addAtHead(2)
    linked_list.addAtHead(1)
    print(f"添加头部元素后: {linked_list}")
    
    linked_list.addAtIndex(3, 0)
    print(f"在索引3添加0后: {linked_list}")
    
    linked_list.deleteAtIndex(2)
    print(f"删除索引2后: {linked_list}")
    
    linked_list.addAtHead(6)
    print(f"添加头部6后: {linked_list}")
    
    linked_list.addAtTail(4)
    print(f"添加尾部4后: {linked_list}")
    
    print(f"获取索引5: {linked_list.get(5)}")  # -1
    print(f"获取索引3: {linked_list.get(3)}")  # 0
    print(f"获取索引2: {linked_list.get(2)}")  # 2
    print(f"获取索引0: {linked_list.get(0)}")  # 6
    
    linked_list.deleteAtIndex(0)
    print(f"删除索引0后，获取索引0: {linked_list.get(0)}")  # 1
    
    # 测试列表实现
    print("\n--- 列表实现测试 ---")
    list_impl = ListLinkedList()
    list_impl.addAtHead(1)
    list_impl.addAtTail(3)
    list_impl.addAtIndex(1, 2)
    print(f"列表实现操作后: {list_impl}")
    print(f"获取索引1: {list_impl.get(1)}")  # 2
    
    list_impl.deleteAtIndex(1)
    print(f"删除索引1后: {list_impl}")
    print(f"获取索引1: {list_impl.get(1)}")  # 3
    
    # 测试deque实现
    print("\n--- Deque实现测试 ---")
    deque_impl = DequeLinkedList()
    deque_impl.addAtHead(1)
    deque_impl.addAtTail(3)
    deque_impl.addAtIndex(1, 2)
    print(f"Deque实现操作后: {deque_impl}")
    print(f"获取索引1: {deque_impl.get(1)}")  # 2
    
    # 测试单链表实现
    print("\n--- 单链表实现测试 ---")
    singly_list = SinglyLinkedList()
    singly_list.addAtHead(1)
    singly_list.addAtTail(3)
    singly_list.addAtIndex(1, 2)
    print(f"单链表实现操作后: [1, 2, 3]")
    print(f"获取索引1: {singly_list.get(1)}")  # 2
    
    # 测试函数式实现
    print("\n--- 函数式实现测试 ---")
    func_list = FunctionalLinkedList()
    func_list.addAtHead(1)
    func_list.addAtTail(3)
    func_list.addAtIndex(1, 2)
    print(f"函数式实现操作后: {func_list}")
    print(f"获取索引1: {func_list.get(1)}")  # 2
    
    # 测试泛型实现
    print("\n--- 泛型实现测试 ---")
    str_list = GenericLinkedList(str)
    str_list.add(0, "Hello")
    str_list.add(1, "World")
    print(f"泛型实现(字符串): {str_list}")
    print(f"获取索引0: {str_list.get(0)}")  # Hello
    
    # 测试函数式操作
    print("\n--- 函数式操作测试 ---")
    operations = [
        ("addAtHead", 7),
        ("addAtHead", 2),
        ("addAtHead", 1),
        ("addAtIndex", 3, 0),
        ("deleteAtIndex", 2),
        ("addAtHead", 6),
        ("addAtTail", 4),
        ("get", 4),
        ("addAtHead", 4),
        ("addAtIndex", 5, 0),
        ("addAtHead", 6),
    ]
    
    results = linked_list_operations(operations)
    print(f"函数式操作结果: {results}")
    
    # 性能比较
    print("\n--- 性能特性比较 ---")
    print("双向链表: 插入删除O(1)，查找O(n)")
    print("列表实现: 尾部操作O(1)，其他O(n)")
    print("Deque实现: 两端操作O(1)，中间O(n)")
    print("单链表: 插入删除O(1)，查找O(n)")
    
    print("\n=== Python特性说明 ===")
    print("1. 动态类型: 运行时类型检查")
    print("2. 自动垃圾回收: 无需手动内存管理")
    print("3. 列表和deque: 高效的内置数据结构")
    print("4. 类型提示: 增强代码可读性")
    print("5. 魔术方法: __str__, __len__等特殊方法")
    print("6. 函数式编程: 支持函数式操作")

if __name__ == "__main__":
    test_design_linked_list()

"""
Python实现特点:

1. **动态类型系统**:
   - 运行时类型检查
   - 类型提示增强可读性
   - 灵活的数据处理

2. **自动内存管理**:
   - 垃圾回收机制
   - 引用计数
   - 无需手动释放内存

3. **丰富的内置数据结构**:
   - list动态数组
   - deque双端队列
   - 高效的操作方法

4. **面向对象特性**:
   - 类和继承
   - 魔术方法
   - 属性装饰器

5. **函数式编程**:
   - 高阶函数
   - Lambda表达式
   - 列表推导式

6. **代码简洁性**:
   - 简洁的语法
   - 强大的标准库
   - 易于阅读和维护

算法复杂度:
- get: O(n)
- addAtHead: O(1)
- addAtTail: O(1)
- addAtIndex: O(n)
- deleteAtIndex: O(n)
- 空间复杂度: O(n)
"""

File: lc_java/lcJava/design/DesignLinkedList.java
// LeetCode 707: 设计链表
// 设计链表的实现。您可以选择使用单链表或双链表。

import java.util.*;
import java.util.function.*;

// 双向链表节点
class ListNode {
    int val;
    ListNode prev;
    ListNode next;
    
    public ListNode() {
        this(0);
    }
    
    public ListNode(int val) {
        this.val = val;
        this.prev = null;
        this.next = null;
    }
    
    @Override
    public String toString() {
        return "ListNode(" + val + ")";
    }
}

// 双向链表实现
class MyLinkedList {
    private ListNode head;
    private ListNode tail;
    private int size;
    
    public MyLinkedList() {
        // 创建哨兵节点
        head = new ListNode();
        tail = new ListNode();
        head.next = tail;
        tail.prev = head;
        size = 0;
    }
    
    public int get(int index) {
        if (index < 0 || index >= size) {
            return -1;
        }
        
        ListNode current = head.next;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current.val;
    }
    
    public void addAtHead(int val) {
        addAtIndex(0, val);
    }
    
    public void addAtTail(int val) {
        addAtIndex(size, val);
    }
    
    public void addAtIndex(int index, int val) {
        if (index < 0 || index > size) {
            return;
        }
        
        ListNode pred = head;
        for (int i = 0; i < index; i++) {
            pred = pred.next;
        }
        
        ListNode succ = pred.next;
        ListNode newNode = new ListNode(val);
        
        newNode.prev = pred;
        newNode.next = succ;
        pred.next = newNode;
        succ.prev = newNode;
        
        size++;
    }
    
    public void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }
        
        ListNode pred = head;
        for (int i = 0; i < index; i++) {
            pred = pred.next;
        }
        
        ListNode toDelete = pred.next;
        ListNode succ = toDelete.next;
        
        pred.next = succ;
        succ.prev = pred;
        
        size--;
    }
    
    public int getSize() {
        return size;
    }
    
    public List<Integer> toList() {
        List<Integer> result = new ArrayList<>();
        ListNode current = head.next;
        while (current != tail) {
            result.add(current.val);
            current = current.next;
        }
        return result;
    }
    
    @Override
    public String toString() {
        return toList().toString();
    }
}

// 使用ArrayList实现的链表
class ArrayLinkedList {
    private List<Integer> data;
    
    public ArrayLinkedList() {
        data = new ArrayList<>();
    }
    
    public int get(int index) {
        if (index >= 0 && index < data.size()) {
            return data.get(index);
        }
        return -1;
    }
    
    public void addAtHead(int val) {
        data.add(0, val);
    }
    
    public void addAtTail(int val) {
        data.add(val);
    }
    
    public void addAtIndex(int index, int val) {
        if (index >= 0 && index <= data.size()) {
            data.add(index, val);
        }
    }
    
    public void deleteAtIndex(int index) {
        if (index >= 0 && index < data.size()) {
            data.remove(index);
        }
    }
    
    @Override
    public String toString() {
        return data.toString();
    }
}

// 使用LinkedList实现
class DequeLinkedList {
    private LinkedList<Integer> data;
    
    public DequeLinkedList() {
        data = new LinkedList<>();
    }
    
    public int get(int index) {
        if (index >= 0 && index < data.size()) {
            return data.get(index);
        }
        return -1;
    }
    
    public void addAtHead(int val) {
        data.addFirst(val);
    }
    
    public void addAtTail(int val) {
        data.addLast(val);
    }
    
    public void addAtIndex(int index, int val) {
        if (index >= 0 && index <= data.size()) {
            data.add(index, val);
        }
    }
    
    public void deleteAtIndex(int index) {
        if (index >= 0 && index < data.size()) {
            data.remove(index);
        }
    }
    
    @Override
    public String toString() {
        return data.toString();
    }
}

// 单链表节点
class SinglyListNode {
    int val;
    SinglyListNode next;
    
    public SinglyListNode() {
        this(0);
    }
    
    public SinglyListNode(int val) {
        this.val = val;
        this.next = null;
    }
}

// 单链表实现
class SinglyLinkedList {
    private SinglyListNode head;
    private int size;
    
    public SinglyLinkedList() {
        head = new SinglyListNode(); // 哨兵节点
        size = 0;
    }
    
    public int get(int index) {
        if (index < 0 || index >= size) {
            return -1;
        }
        
        SinglyListNode current = head.next;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current.val;
    }
    
    public void addAtHead(int val) {
        addAtIndex(0, val);
    }
    
    public void addAtTail(int val) {
        addAtIndex(size, val);
    }
    
    public void addAtIndex(int index, int val) {
        if (index < 0 || index > size) {
            return;
        }
        
        SinglyListNode pred = head;
        for (int i = 0; i < index; i++) {
            pred = pred.next;
        }
        
        SinglyListNode newNode = new SinglyListNode(val);
        newNode.next = pred.next;
        pred.next = newNode;
        
        size++;
    }
    
    public void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }
        
        SinglyListNode pred = head;
        for (int i = 0; i < index; i++) {
            pred = pred.next;
        }
        
        pred.next = pred.next.next;
        size--;
    }
}

// 泛型链表实现
class GenericLinkedList<T> {
    private static class Node<T> {
        T val;
        Node<T> prev;
        Node<T> next;
        
        public Node(T val) {
            this.val = val;
        }
    }
    
    private Node<T> head;
    private Node<T> tail;
    private int size;
    
    public GenericLinkedList() {
        head = new Node<>(null);
        tail = new Node<>(null);
        head.next = tail;
        tail.prev = head;
        size = 0;
    }
    
    public T get(int index) {
        if (index < 0 || index >= size) {
            return null;
        }
        
        Node<T> current = head.next;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current.val;
    }
    
    public void add(int index, T val) {
        if (index < 0 || index > size) {
            return;
        }
        
        Node<T> pred = head;
        for (int i = 0; i < index; i++) {
            pred = pred.next;
        }
        
        Node<T> succ = pred.next;
        Node<T> newNode = new Node<>(val);
        
        newNode.prev = pred;
        newNode.next = succ;
        pred.next = newNode;
        succ.prev = newNode;
        
        size++;
    }
    
    public void delete(int index) {
        if (index < 0 || index >= size) {
            return;
        }
        
        Node<T> pred = head;
        for (int i = 0; i < index; i++) {
            pred = pred.next;
        }
        
        Node<T> toDelete = pred.next;
        Node<T> succ = toDelete.next;
        
        pred.next = succ;
        succ.prev = pred;
        
        size--;
    }
    
    public List<T> toList() {
        List<T> result = new ArrayList<>();
        Node<T> current = head.next;
        while (current != tail) {
            result.add(current.val);
            current = current.next;
        }
        return result;
    }
}

// 函数式编程风格的链表操作
class FunctionalLinkedList {
    private List<Integer> data;
    
    public FunctionalLinkedList() {
        data = new ArrayList<>();
    }
    
    public int get(int index) {
        return Optional.of(index)
                .filter(i -> i >= 0 && i < data.size())
                .map(data::get)
                .orElse(-1);
    }
    
    public void addAtHead(int val) {
        data.add(0, val);
    }
    
    public void addAtTail(int val) {
        data.add(val);
    }
    
    public void addAtIndex(int index, int val) {
        Optional.of(index)
                .filter(i -> i >= 0 && i <= data.size())
                .ifPresent(i -> data.add(i, val));
    }
    
    public void deleteAtIndex(int index) {
        Optional.of(index)
                .filter(i -> i >= 0 && i < data.size())
                .ifPresent(data::remove);
    }
    
    @Override
    public String toString() {
        return data.toString();
    }
}

// 链表操作接口
interface LinkedListOperations {
    int get(int index);
    void addAtHead(int val);
    void addAtTail(int val);
    void addAtIndex(int index, int val);
    void deleteAtIndex(int index);
}

// 抽象链表类
abstract class AbstractLinkedList implements LinkedListOperations {
    protected int size;
    
    public int getSize() {
        return size;
    }
    
    protected boolean isValidIndex(int index) {
        return index >= 0 && index < size;
    }
    
    protected boolean isValidInsertIndex(int index) {
        return index >= 0 && index <= size;
    }
}

// 具体实现类
class ConcreteLinkedList extends AbstractLinkedList {
    private List<Integer> data;
    
    public ConcreteLinkedList() {
        data = new ArrayList<>();
        size = 0;
    }
    
    @Override
    public int get(int index) {
        return isValidIndex(index) ? data.get(index) : -1;
    }
    
    @Override
    public void addAtHead(int val) {
        addAtIndex(0, val);
    }
    
    @Override
    public void addAtTail(int val) {
        addAtIndex(size, val);
    }
    
    @Override
    public void addAtIndex(int index, int val) {
        if (isValidInsertIndex(index)) {
            data.add(index, val);
            size++;
        }
    }
    
    @Override
    public void deleteAtIndex(int index) {
        if (isValidIndex(index)) {
            data.remove(index);
            size--;
        }
    }
}

public class DesignLinkedList {
    
    // 函数式操作链表
    public static List<Integer> linkedListOperations(List<String[]> operations) {
        MyLinkedList obj = new MyLinkedList();
        List<Integer> results = new ArrayList<>();
        
        for (String[] operation : operations) {
            String op = operation[0];
            switch (op) {
                case "get":
                    results.add(obj.get(Integer.parseInt(operation[1])));
                    break;
                case "addAtHead":
                    obj.addAtHead(Integer.parseInt(operation[1]));
                    results.add(null);
                    break;
                case "addAtTail":
                    obj.addAtTail(Integer.parseInt(operation[1]));
                    results.add(null);
                    break;
                case "addAtIndex":
                    obj.addAtIndex(Integer.parseInt(operation[1]), Integer.parseInt(operation[2]));
                    results.add(null);
                    break;
                case "deleteAtIndex":
                    obj.deleteAtIndex(Integer.parseInt(operation[1]));
                    results.add(null);
                    break;
                default:
                    results.add(null);
            }
        }
        
        return results;
    }
    
    public static void testDesignLinkedList() {
        System.out.println("=== 设计链表测试 ===");
        
        // 测试双向链表实现
        System.out.println("\n--- 双向链表实现测试 ---");
        MyLinkedList list = new MyLinkedList();
        
        list.addAtHead(7);
        list.addAtHead(2);
        list.addAtHead(1);
        System.out.println("添加头部元素后: " + list);
        
        list.addAtIndex(3, 0);
        System.out.println("在索引3添加0后: " + list);
        
        list.deleteAtIndex(2);
        System.out.println("删除索引2后: " + list);
        
        list.addAtHead(6);
        System.out.println("添加头部6后: " + list);
        
        list.addAtTail(4);
        System.out.println("添加尾部4后: " + list);
        
        System.out.println("获取索引5: " + list.get(5)); // -1
        System.out.println("获取索引3: " + list.get(3)); // 0
        System.out.println("获取索引2: " + list.get(2)); // 2
        System.out.println("获取索引0: " + list.get(0)); // 6
        
        list.deleteAtIndex(0);
        System.out.println("删除索引0后，获取索引0: " + list.get(0)); // 1
        
        // 测试ArrayList实现
        System.out.println("\n--- ArrayList实现测试 ---");
        ArrayLinkedList arrayList = new ArrayLinkedList();
        arrayList.addAtHead(1);
        arrayList.addAtTail(3);
        arrayList.addAtIndex(1, 2);
        System.out.println("ArrayList实现操作后: " + arrayList);
        System.out.println("获取索引1: " + arrayList.get(1)); // 2
        
        arrayList.deleteAtIndex(1);
        System.out.println("删除索引1后: " + arrayList);
        System.out.println("获取索引1: " + arrayList.get(1)); // 3
        
        // 测试LinkedList实现
        System.out.println("\n--- LinkedList实现测试 ---");
        DequeLinkedList dequeList = new DequeLinkedList();
        dequeList.addAtHead(1);
        dequeList.addAtTail(3);
        dequeList.addAtIndex(1, 2);
        System.out.println("LinkedList实现操作后: " + dequeList);
        System.out.println("获取索引1: " + dequeList.get(1)); // 2
        
        // 测试单链表实现
        System.out.println("\n--- 单链表实现测试 ---");
        SinglyLinkedList singlyList = new SinglyLinkedList();
        singlyList.addAtHead(1);
        singlyList.addAtTail(3);
        singlyList.addAtIndex(1, 2);
        System.out.println("单链表实现操作后: [1, 2, 3]");
        System.out.println("获取索引1: " + singlyList.get(1)); // 2
        
        // 测试泛型实现
        System.out.println("\n--- 泛型实现测试 ---");
        GenericLinkedList<String> strList = new GenericLinkedList<>();
        strList.add(0, "Hello");
        strList.add(1, "World");
        System.out.println("泛型实现(字符串): " + strList.toList());
        System.out.println("获取索引0: " + strList.get(0)); // Hello
        
        // 测试函数式实现
        System.out.println("\n--- 函数式实现测试 ---");
        FunctionalLinkedList funcList = new FunctionalLinkedList();
        funcList.addAtHead(1);
        funcList.addAtTail(3);
        funcList.addAtIndex(1, 2);
        System.out.println("函数式实现操作后: " + funcList);
        System.out.println("获取索引1: " + funcList.get(1)); // 2
        
        // 测试抽象类实现
        System.out.println("\n--- 抽象类实现测试 ---");
        ConcreteLinkedList concreteList = new ConcreteLinkedList();
        concreteList.addAtHead(1);
        concreteList.addAtTail(3);
        concreteList.addAtIndex(1, 2);
        System.out.println("抽象类实现大小: " + concreteList.getSize());
        System.out.println("获取索引1: " + concreteList.get(1)); // 2
        
        // 测试函数式操作
        System.out.println("\n--- 函数式操作测试 ---");
        List<String[]> operations = Arrays.asList(
            new String[]{"addAtHead", "7"},
            new String[]{"addAtHead", "2"},
            new String[]{"addAtHead", "1"},
            new String[]{"addAtIndex", "3", "0"},
            new String[]{"deleteAtIndex", "2"},
            new String[]{"addAtHead", "6"},
            new String[]{"addAtTail", "4"},
            new String[]{"get", "4"}
        );
        
        List<Integer> results = linkedListOperations(operations);
        System.out.println("函数式操作结果: " + results);
        
        System.out.println("\n=== Java特性说明 ===");
        System.out.println("1. 面向对象: 类、继承、多态、封装");
        System.out.println("2. 泛型: 类型安全的集合和方法");
        System.out.println("3. 集合框架: ArrayList、LinkedList等");
        System.out.println("4. 接口和抽象类: 契约定义和部分实现");
        System.out.println("5. Lambda表达式: 函数式编程支持");
        System.out.println("6. Optional: 空值安全处理");
    }
    
    public static void main(String[] args) {
        testDesignLinkedList();
    }
}

/*
Java实现特点:

1. **面向对象编程**:
   - 类和对象封装
   - 继承和多态
   - 接口契约定义

2. **泛型编程**:
   - 类型安全
   - 编译时类型检查
   - 代码复用

3. **集合框架**:
   - ArrayList动态数组
   - LinkedList双向链表
   - 丰富的操作方法

4. **内存管理**:
   - 自动垃圾回收
   - 引用管理
   - 内存泄漏预防

5. **函数式编程**:
   - Lambda表达式
   - Stream API
   - Optional类

6. **异常处理**:
   - 编译时异常检查
   - try-catch机制
   - 资源自动管理

算法复杂度:
- get: O(n)
- addAtHead: O(1)
- addAtTail: O(1)
- addAtIndex: O(n)
- deleteAtIndex: O(n)
- 空间复杂度: O(n)
*/